conwayste/src/config.rs

```diff
@@ -18,373 +18,607 @@
 
 extern crate toml;
 
+use crate::constants::{CONFIG_FILE_PATH, DEFAULT_ZOOM_LEVEL, MIN_CONFIG_FLUSH_TIME};
+use std::error::Error;
+use std::fmt;
+use std::ops::{AddAssign, SubAssign};
+use std::time::{Duration, Instant};
+
 use std::fs::OpenOptions;
-use std::io::{Write, Read};
+use std::io::Read;
+#[cfg(not(test))]
+use std::io::Write;
+#[cfg(not(test))]
 use std::path::Path;
-use crate::constants::{CONFIG_FILE_PATH, DEFAULT_ZOOM_LEVEL};
 
-/// User configuration descriptor which contains all of the configurable settings
-/// for this game. These *should* be modified within the game, but one can 
-/// always edit this file directly. The game will fail to load if there are
-/// any errors parsing the `conwayste.toml` file.
+type TomlMap = toml::map::Map<String, toml::Value>;
+use toml::Value;
+
+#[derive(Debug)]
+pub struct ConfigError {
+    pub msg: String,
+}
+
+impl fmt::Display for ConfigError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        write!(f, "{:?}", self)?;
+        Ok(())
+    }
+}
+
+impl Error for ConfigError {}
+
+fn new_config_error(msg: String) -> Box<dyn Error> {
+    Box::new(ConfigError { msg })
+}
+
+lazy_static! {
+    /// The default configuration, in TOML format.
+    static ref DEFAULT_STRING: String = {
+        let default_settings: Settings = Default::default();
+        toml::to_string(&default_settings).unwrap()
+    };
+
+    /// A TomlMap for the `DEFAULT_STRING`.
+    static ref DEFAULT_MAP: TomlMap = toml::from_str(DEFAULT_STRING.as_str()).unwrap();
+
+    /// Same as `DEFAULT_STRING` but as a TOML comment and with version string at top.
+    static ref COMMENTED_DEFAULT_STRING: String = {
+        let mut s = String::new();
+        //XXX version string
```

Comment:
Was this a to-do reminder? The `XXX`, since it looks like you do have that below via `version!()`.

Reply:
Oops, yeah, forgot to remove.